backupctl.8.rst
===========
 backupctl
===========

------------------------------------------------------------
 Manage dirvish backups with an underlying zfs storage pool
------------------------------------------------------------

:Author:
    Written by Tobias Rueetschi, Adfinis SyGroup AG.
:Date:
    May 2018
:Copyright:
    Copyright 2018 Adfinis SyGroup AG License GPLv3+:
    GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
    This is free software: you are free to change and redistribute it.
    There is NO WARRANTY, to the extent permitted by law.
:Version:
    2.0.5
:Manual section:
    8
:Manual group:
    System Manager's Manual


SYNOPSIS
=========

backupctl COMMAND [OPTIONS]

backupctl log

backupctl [--help]


DESCRIPTION
============
Tool to manage zfs volumes and create new dirvish vault configurations. This
tool depends on a already present zfs pool, normally called backup.
This tool (which is pronounced "backup cuddle") should be run with root
privileges, if not possible, at least zfs fuse access as user is needed
(to create new or remove existing zfs datasets and set quotas).
This is more tricky and may not be tested as well as zfs with the kernel
module.


COMMANDS
=========
All subcommands that modify zfs volumes and dirvish configurations.

new -n customer [-v server/vault] [-s size]
--------------------------------------------
Create a new customer zfs volume or a new dirvish vault inside a customer.

resize -n customer [-v server/vault] -s size
---------------------------------------------
Resize an existing customer zfs volume or dirvish vault inside a customer.
Shrinking is supported too, but if there is more data in the vault than the
size you're trying to shrink to an error will occur.
To unset a quota, set the parameter size to "none".

remove -n customer [-v server/vault]
-------------------------------------
Remove an existing customer zfs volume or dirvish vault inside a customer.
If removing a customer, all dirvish vaults inside this customer will be removed
too.

log
----
Show the history of generating, resizing and removing customers and servers.


OPTIONS
========

-v, --vault             Specify the hostname of the dirvish vault name.
--dirvish-client        Specify a client ip or fqdn to back up. Needed if
                        different from option client.
-h, --help              Show this help message and exit.
-n, --name              Customer name. Used as a logical group of dirvish
                        vaults.
-s, --size              Quota of a customer or a server. Size can be written
                        human readable as MB, GB and so on.


QUOTA
======
The storage used in a vault is substracted from the quota of the customer.
The quota of the customer should be equal or bigger than the one of the
vault/server, if this isn't the case, then the quota of the customer can't be
used in the full size. A quota must be defined for a customer. It may be
defined for a vault/server.
The summary of all quotas of vaults inside a customer can be bigger than the
quota of the customer.

Example
--------
As an example, if the ZFS pool is 50 GB, there is one customer with 40 GB, and
there are three vaults/servers with each 20 GB, each vault/server can use up to
20 GB, if all three vaults/servers are not more than 40 GB.

.. code-block::

  NAME                               USED  AVAIL  REFER
  backup                              35G    50G   205K
  backup/customer1                    35G    40G   273K
  backup/customer1/www1.example.com   10G    20G    10G
  backup/customer1/www2.example.com   10G    20G    10G
  backup/customer1/www3.example.com   15G    20G    15G


EXAMPLES
=========

CUSTOMER MANAGEMENT
--------------------

Create a new customer with a quota of 10G.

  backupctl new -n customer1 -s 10G

Resize the quota of an existing customer to 20G.

  backupctl resize -n customer1 -s 20G

Remove a customer with all his servers. All data will be lost.

  backupctl remove -n customer1

SERVER MANAGEMENT
------------------

Create a new server/vault www.example.com for customer1 with no special quota
(it will inherit the quota of the customer).

  backupctl new -n customer1 -v www.example.com

Create a new server/vault www.example.com for customer1 with a quota of 10G.
Data inside this server will also count on the customer level).

  backupctl new -n customer1 -v www.example.com -s 10G

Create a new server/vault www.example.com which isn't directly accessible, so a
special dirvish client is needed, in this case the server is reachable with the
ip address 192.0.2.100.

  backupctl new -n customer1 -v www.example.com -s 10G --dirvish-client 192.0.2.100

Resize the quota of the server/vault www.example.com of customer1 to the size
of 20G.

  backupctl resize -n customer1 -v www.example.com -s 20G

Remove the server/vault www.example.com of customer1, this will delete all the
backups and also the dirvish configuration for this server.

  backupctl remove -n customer1 -v www.example.com


EXIT STATUS
============
The following exit values are returned:

0
--
Successful completion.

1
--
An error occurred.

2
--
No command was given.


FILES
======

/etc/backupctl.ini
-------------------
System-wide configuration file.

$XDG_CONFIG_HOME/backupctl.ini
-------------------------------
User specific configuration file. Will overwrite the configuration options of
the system-wide configuration file.

$PWD/backupctl.ini
-------------------
Local configuration file. Will overwrite the configuration options of any
previous configuration file.

/var/lib/backupctl/backupctl.db
--------------------------------
backupctl history database. This is an sqlite3 database.


SEE ALSO
=========

* ``backupctl.ini(5)``, backupctl configuration file
* ``dirvish(8)``, dirvish backup utility

.. vim: set et ts=2 sw=2 :


backupctl/backupctl.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse
import configparser
import logging
import os
import sys

import sqlalchemy
from xdg import BaseDirectory

from backupctl import zfs
from backupctl.dirvish import Dirvish
from backupctl.history import History
from backupctl.version import __version__

LOG = logging.getLogger(__name__)
LOG.addHandler(logging.StreamHandler())
LOG.setLevel(logging.WARN)


def main():
    """Main function.
    Manage cli arguments and call the corresponding function per subcommand
    with the needed arguments.
    """
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description="""\
        Manage dirvish backups with an underlying zfs storage pool.
        """,
    )
    parser.add_argument(
        "command",
        help="""\
        All subcommands that modify zfs file system and dirvish configurations.
        See the manpage (man 8 backupctl) for more inforation.
        """,
        choices=["new", "resize", "remove", "log"]
    )
    parser.add_argument(
        "-v",
        "--vault",
        required=False,
        default=None,
        help="""\
        Dirvish vault name or server hostname to backup.
        Use a fully qualified domain name for that.
        """,
    )
    parser.add_argument(
        "--dirvish-client",
        required=False,
        default=None,
        help="""\
        Specify an ip address or fqdn for a dirvish vault to back up.
        Needed if different from option vault.
        """,
    )
    parser.add_argument(
        "-n",
        "--customer",
        required=False,
        default=None,
        help="""\
        Customer name. Used as a logical group of dirvish vaults.
        """,
    )
    parser.add_argument(
        "-s",
        "--size",
        required=False,
        default=None,
        help="""\
        Quota of a customer or a vault. Size can be written human readable as
        MB, GB and so on.
        """,
    )
    parser.add_argument(
        "--version",
        action="version",
        version="Backupctl v{0}".format(__version__),
        help="""\
        Displays the backupctl version
        """,
    )
    args = parser.parse_args()

    cfg = config()

    if not os.path.exists(os.path.dirname(cfg["database"].get("path"))):
        os.makedirs(os.path.dirname(cfg["database"].get("path")))

    try:
        engine = sqlalchemy.create_engine(cfg["database"].get("fullpath"))
        LOG.debug(
            "Opened database {0} successfully".format(cfg["database"].get("fullpath"))
        )
    except sqlalchemy.exc.ArgumentError as e:
        LOG.error(
            "Couldn't open database {0}. Exit now.".format(
                cfg["database"].get("fullpath")
            )
        )
        sys.exit(1)
    except sqlalchemy.exc.OperationalError as e:
        LOG.error(
            "Couldn't open database {0}. Exit now.".format(
                cfg["database"].get("fullpath")
            )
        )
        sys.exit(1)

    hist = History(engine)
    dirvish = Dirvish(engine)

    if args.command == "new":
        try:
            new(
                hist,
                dirvish,
                cfg["zfs"]["pool"],
                cfg["zfs"]["root"],
                args.customer,
                args.vault,
                args.size,
                args.dirvish_client,
            )
        except KeyError as e:
            LOG.error(
                "ZFS Pool and ZFS Root must be specified in the "
                "configuration file. Exit now."
            )
            sys.exit(1)
    elif args.command == "resize":
        try:
            resize(hist, cfg["zfs"]["pool"], args.customer, args.vault, args.size)
        except KeyError as e:
            LOG.error(
                "ZFS Pool must be specified in the configuration file. Exit now."
            )
            sys.exit(1)
    elif args.command == "remove":
        try:
            remove(hist, cfg["zfs"]["pool"], args.customer, args.vault)
        except KeyError as e:
            LOG.error(
                "ZFS Pool must be specified in the configuration file. Exit now."
            )
            sys.exit(1)
    elif args.command == "log":
        history_show(hist)
    else:
        sys.exit(1)
    sys.exit(0)


def backup_start():
    """Add an entry to the database when a dirvish backup is started.

    This function should be triggered by dirvish pre-server.
    """
    cfg = config()
    if not os.path.exists(os.path.dirname(cfg["database"].get("path"))):
        os.makedirs(os.path.dirname(cfg["database"].get("path")))
    try:
        engine = sqlalchemy.create_engine(cfg["database"].get("fullpath"))
        LOG.debug(
            "Opened database {0} successfully".format(cfg["database"].get("fullpath"))
        )
    except (sqlalchemy.exc.ArgumentError, sqlalchemy.exc.OperationalError) as e:
        LOG.error(
            "Couldn't open database {0}. Exit now.".format(
                cfg["database"].get("fullpath")
            )
        )
        sys.exit(1)
    dirvish = Dirvish(engine)
    dirvish.backup_start()


def backup_stop():
    """Add an entry to the database when a dirvish backup is stopped.

    This function should be triggered by dirvish post-server.
    """
    cfg = config()
    if not os.path.exists(os.path.dirname(cfg["database"].get("path"))):
        os.makedirs(os.path.dirname(cfg["database"].get("path")))
    try:
        engine = sqlalchemy.create_engine(cfg["database"].get("fullpath"))
        LOG.debug(
            "Opened database {0} successfully".format(cfg["database"].get("fullpath"))
        )
    except sqlalchemy.exc.ArgumentError as e:
        LOG.error(
            "Couldn't open database {0}. Exit now.".format(
                cfg["database"].get("fullpath")
            )
        )
        sys.exit(1)
    except sqlalchemy.exc.OperationalError as e:
        LOG.error(
            "Couldn't open database {0}. Exit now.".format(
                cfg["database"].get("fullpath")
            )
        )
        sys.exit(1)
    dirvish = Dirvish(engine)
    dirvish.backup_stop()


def config():
    """Read the configuration files. If no configuration exists, write the
    default configuration to the directory ~/.config.

    :returns: Configuration object.
    :rtype: `configparser.ConfigParser`
    """
    cfg = configparser.ConfigParser()
    cfg.read(os.path.join(os.sep, "etc", "backupctl.ini"))
    cfg.read(os.path.join(BaseDirectory.xdg_config_home, "backupctl.ini"))
    cfg.read("backupctl.ini")

    if not cfg.has_section("database"):
        cfg.add_section("database")
    if not cfg.has_option("database", "type"):
        cfg["database"]["type"] = "sqlite"
    if not cfg.has_option("database", "path"):
        cfg["database"]["path"] = os.path.join(
            os.sep, "var", "lib", "backupctl", "backupctl.db"
        )
    if not cfg.has_option("database", "fullpath"):
        cfg["database"]["fullpath"] = "{0}:///{1}".format(
            cfg["database"].get("type"), cfg["database"].get("path")
        )
    return cfg


def new(hist, dirvish, pool, root, customer, vault=None, size=None, client=None):
    """Create a new customer or a new vault/server.

    :param history.History hist:    History database.
    :param dirvish.Dirvish dirvish: Dirvish object.
    :param string pool:             ZFS pool name.
    :param string root:             Backup root path.
    :param string customer:         Customer name.
    :param string vault:            Vault name or server hostname.
    :param string size:             Quota for this customer or vault.
    """
    if not customer:
        LOG.error("Customer is needed")
        sys.exit(1)
    if not vault and not size:
        LOG.error("If no vault is given, a size is required")
        sys.exit(1)
    if vault is not None:
        fs = os.path.join(pool, customer, vault)
        path = os.path.join(root, customer, vault)
    else:
        fs = os.path.join(pool, customer)
        path = None
    fs_status = zfs.new_filesystem(fs, path, size)
    if fs_status:
        hist.add(customer, "create", vault, size)
        if vault is not None:
            if client is None:
                client = vault
            dirvish.create_config(root, customer, vault, client)
            hist.add(customer, "config", vault)


def resize(hist, pool, customer, vault=None, size=None):
    """Resize an existing customer or vault.

    :param history.History hist:    History database.
    :param string pool:             ZFS pool name.
    :param string customer:         Customer name.
    :param string vault:            Vault name or server hostname.
    :param string size:             Quota for this customer or vault.
    """
    if not customer:
        LOG.error("Customer is needed")
        sys.exit(1)
    if not size:
        LOG.error("A size is required")
        sys.exit(1)
    if vault:
        fs = os.path.join(pool, customer, vault)
    else:
        fs = os.path.join(pool, customer)
    zfs.resize_filesystem(fs, size)
    hist.add(customer, "resize", vault, size)


def remove(hist, pool, customer, vault=None):
    """Remove a customer or vault.

    :param history.History hist:    History database.
    :param string pool:             ZFS pool name.
    :param string customer:         Customer name.
    :param string vault:            Vault name or server hostname.
    """
    if not customer:
        LOG.error("Customer is needed")
        sys.exit(1)
    if vault:
        fs = os.path.join(pool, customer, vault)
    else:
        fs = os.path.join(pool, customer)
    zfs.remove_filesystem(fs)
    if os.path.isdir(fs):
        os.rmdir(fs)
    hist.add(customer, "remove", vault)


def history_show(history):
    """Manage the command log.

    :param history.History history: History database.
    """
    for row in history.show():
        print(row)


if __name__ == "__main__":
    main()
